extensions/io/org.eclipse.smarthome.io.mqttembeddedbroker/src/main/java/org/eclipse/smarthome/io/mqttembeddedbroker/internal/EmbeddedBrokerServiceImpl.java

```diff
@@ -0,0 +1,304 @@
+ */
+@Component(immediate = true, service = EmbeddedBrokerService.class, configurationPid = "org.eclipse.smarthome.mqttembeddedbroker", property = {
+        org.osgi.framework.Constants.SERVICE_PID + "=org.eclipse.smarthome.mqttembeddedbroker",
+        ConfigurableService.SERVICE_PROPERTY_DESCRIPTION_URI + "=mqtt:mqttembeddedbroker",
+        ConfigurableService.SERVICE_PROPERTY_CATEGORY + "=MQTT",
+        ConfigurableService.SERVICE_PROPERTY_LABEL + "=MQTT Embedded Broker" })
+public class EmbeddedBrokerServiceImpl implements EmbeddedBrokerService, ConfigurableService, MqttConnectionObserver,
+        MqttServiceObserver, MqttEmbeddedBrokerStartedListener, BrokerMetricsListener {
+    private MqttService service;
+    // private NetworkServerTls networkServerTls; //TODO wait for NetworkServerTls implementation
+
+    protected Server server;
+    private final Logger logger = LoggerFactory.getLogger(EmbeddedBrokerServiceImpl.class);
+    protected MqttEmbeddedBrokerDetectStart detectStart = new MqttEmbeddedBrokerDetectStart(this);
+    protected MqttEmbeddedBrokerMetrics metrics = new MqttEmbeddedBrokerMetrics(this);
+
+    private MqttBrokerConnection connection;
+
+    @Reference(cardinality = ReferenceCardinality.MANDATORY)
+    public void setMqttService(MqttService service) {
+        this.service = service;
+    }
+
+    public void unsetMqttService(MqttService service) {
+        this.service = service;
+    }
+
+    // TODO wait for NetworkServerTls implementation
+    // @Reference(cardinality = ReferenceCardinality.MANDATORY)
+    // public void setNetworkServerTls(NetworkServerTls networkServerTls) {
+    // this.networkServerTls = networkServerTls;
+    // }
+
+    @Activate
+    public void activate(Map<String, Object> data) {
+        initialize(new Configuration(data).as(ServiceConfiguration.class));
+    }
+
+    public void initialize(ServiceConfiguration config) {
+        if (config.port == null) {
+            config.port = config.secure ? 8883 : 1883;
+        }
+
+        // Create MqttBrokerConnection
+        connection = service.getBrokerConnection(Constants.CLIENTID);
+        if (connection != null) {
+            // Close the existing connection and remove it from the service
+            connection.stop();
+            service.removeBrokerConnection(Constants.CLIENTID);
+        }
+
+        connection = new MqttBrokerConnection("127.0.0.1", config.port, config.secure, Constants.CLIENTID);
+        connection.addConnectionObserver(this);
+
+        if (config.username != null) {
+            connection.setCredentials(config.username, config.password);
+        }
+
+        // Start embedded server
+        try {
+            startEmbeddedServer(config.port, config.secure, config.username, config.password, config.persistenceFile);
+        } catch (IOException e) {
+            logger.debug("Could not start embedded broker", e);
+            logger.debug(e.getLocalizedMessage());
+            return;
```

Comment:
Does it make sense to hide the not startable embedded server and keep the OSGi framework informed that the activation has been done successfully? Other services that rely on this one are activated after that. If you throw the exception so the caller of the activate method (the DS implementation) knows that there has been something wrong and will not activate services that depends on this one...

Reply:
I agree